Review followup: PR #20351 — GH#20312: fix prompt-guard staleness check to use deploy stamp mtime

## Unaddressed review bot suggestions

PR #20351 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/20351

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `prompts/build.txt`
section 6a). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `prompts/build.txt` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- `.agents/scripts/security-posture-helper.sh:1003`
- `.agents/scripts/security-posture-helper.sh:1024`

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

### Inline comments

#### gemini-code-assist on `.agents/scripts/security-posture-helper.sh:1003`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/20351#discussion_r3119953120)

```diff
+		local stamp_mtime
+		# shellcheck disable=SC2086
+		stamp_mtime=$(stat $stat_flag "$deployed_sha_file" 2>/dev/null || echo "0")
+		if [[ "$stamp_mtime" -gt 0 ]]; then
+			ref_epoch="$stamp_mtime"
+			ref_source="deploy stamp"
+		fi
+	fi
+
+	# Level 2: upstream git commit date (only if stamp was not found)
+	if [[ "$ref_epoch" -eq 0 ]]; then
+		local framework_repo="${HOME}/Git/aidevops"
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> The path to the framework repository is hardcoded to `~/Git/aidevops`. To make git operations more robust and portable, avoid hardcoded paths and dynamically determine the repository root. This aligns with the repository's practice of explicitly determining git metadata rather than relying on environment defaults or hardcoded assumptions.
> 
> <details>
> <summary>References</summary>
> 
> 1. To make git operations in scripts more robust, dynamically and explicitly determine environment details instead of relying on hardcoded values or the current checkout state.
> </details>

#### gemini-code-assist on `.agents/scripts/security-posture-helper.sh:1024`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/20351#discussion_r3119953127)

```diff
+	fi
+
+	# Level 3: deployed file mtime (fallback — preserves prior behaviour)
+	if [[ "$ref_epoch" -eq 0 ]]; then
+		# shellcheck disable=SC2086
+		ref_epoch=$(stat $stat_flag "$yaml_file" 2>/dev/null || echo "0")
+	fi
+
+	local file_age_days=0
+	if [[ "$ref_epoch" -gt 0 ]]; then
+		file_age_days=$(((now - ref_epoch) / 86400))
 	fi
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> If the fallback chain fails to provide a valid epoch, `file_age_days` might be empty or default to 0. In shell scripts, especially those using `set -e`, ensure that variables used in arithmetic are handled robustly. Use parameter expansion with a default value (e.g., `${var:-0}`) to prevent arithmetic expansion errors if preceding commands like `stat` or `cat` return empty strings or fail.
> 
> <details>
> <summary>References</summary>
> 
> 1. In shell scripts, use parameter expansion with a default value (e.g., $((... - ${var:-0}))) to handle empty strings robustly in arithmetic expansions, preventing failures when commands return empty results.
> </details>

### PR review summaries

#### gemini-code-assist[bot] review summary

[View review](https://github.com/marcusquinn/aidevops/pull/20351#pullrequestreview-4150400580)

> ## Code Review
> 
> This pull request refactors the staleness check for prompt guard patterns in `security-posture-helper.sh` by implementing a three-level fallback chain (deployment stamp, git commit date, and file mtime). This change addresses issues where `rsync` preserves original file timestamps during deployment, leading to false staleness reports. A new test script is also added to verify the fallback logic. Review feedback suggests improving the script's portability by avoiding hardcoded repository paths and increasing robustness by using shell parameter expansion for arithmetic operations.
<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.8.91 automated scan.

--- a/.agents/scripts/security-posture-helper.sh
+++ b/.agents/scripts/security-posture-helper.sh
@@ -1000,7 +1000,7 @@ check_prompt_guard_patterns() {
 
 	# Level 2: upstream git commit date (only if stamp was not found)
 	if [[ "$ref_epoch" -eq 0 ]]; then
-		local framework_repo="${HOME}/Git/aidevops"
+		local framework_repo="${AIDEVOPS_FRAMEWORK_REPO:-${HOME}/Git/aidevops}"
 		if [[ -d "${framework_repo}/.git" ]]; then
 			local git_epoch
 			git_epoch=$(git -C "$framework_repo" log -1 --format=%ct -- \
@@ -1020,7 +1020,7 @@ check_prompt_guard_patterns() {
 
 	local file_age_days=0
 	if [[ "$ref_epoch" -gt 0 ]]; then
-		file_age_days=$(((now - ref_epoch) / 86400))
+		file_age_days=$(( (now - ${ref_epoch:-0}) / 86400 ))
 	fi
 
 	if [[ "$file_age_days" -gt 30 ]]; then

